feat: Implement map image attachment and toggle functionality

ngoiyaeric opened the pull request:
### **User description**
This commit introduces the ability for you to attach an image to be displayed on the map, along with controls to manage its visibility and removal.

Key changes:

- **Image Attachment:**
    - Added a file input to `ChatPanel` allowing you to select image files.
    - Selected images are read as data URLs and stored in `MapDataContext`.
- **Map Rendering:**
    - `MapboxMap` component now consumes the attached image data from `MapDataContext`.
    - Attached images are rendered as a raster layer on the map using a default coordinate quad.
    - The map updates automatically if the image is changed or cleared.
- **Visibility Toggle:**
    - `MapToggleContext` updated to include state for image visibility (`isAttachedImageVisible`).
    - `MapToggle` component now includes a button (Eye/EyeOff icon) to toggle the attached image's visibility. This button only appears if an image is attached.
    - `MapboxMap` respects this toggle, showing or hiding the image layer without removing it from the source unless the image is cleared entirely.
- **Remove Functionality:**
    - Added a "Remove Image" button (XCircle icon) to `ChatPanel`.
    - This button clears the `attachedImage` data from `MapDataContext`, causing the image to be removed from the map.
- **Context Updates:**
    - `MapDataContext` (`map-data-context.tsx`) updated to include `attachedImage?: string | File | null;`.
    - `MapToggleContext` (`map-toggle-context.tsx`) updated to include `isAttachedImageVisible` and its setter.
- **UI Adjustments:**
    - Icons and buttons added to `ChatPanel` and `MapToggle`.
    - Padding adjusted in `ChatPanel` to accommodate new UI elements.
    - Added `title` attributes for better accessibility on new interactive elements.

This feature enhances the map's utility by allowing visual information to be overlaid directly by you.


___

### **PR Type**
Enhancement


___

### **Description**
- Added image attachment and removal functionality to map UI
  - Users can select and attach images to overlay on the map
  - Images can be removed via a dedicated button

- Implemented visibility toggle for attached map images
  - Eye/EyeOff button in map controls toggles image layer visibility

- Updated map rendering logic to display attached images as raster layers
  - Dynamically adds/removes image sources and layers based on context

- Extended context providers to manage attached image state and visibility


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>chat-panel.tsx</strong><dd><code>Enable image attachment and removal in chat panel</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/chat-panel.tsx

<li>Added file input for image attachment to map<br> <li> Added button to remove attached image<br> <li> Adjusted UI padding and button positions for new controls<br> <li> Integrated attached image state with MapDataContext


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/179/files#diff-06a509597829dfc3e720c47b51047e08ba858772f51c5dfd2e01bd6deefb4241">+53/-6</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>map-toggle-context.tsx</strong><dd><code>Add image visibility state to map toggle context</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/map-toggle-context.tsx

<li>Added isAttachedImageVisible state and setter to context<br> <li> Updated provider to include image visibility controls


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/179/files#diff-5298f43fdfb7cb81920be8f46b932adbf8ec5c767e4173ea608bef529c06cc59">+11/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>map-toggle.tsx</strong><dd><code>Add attached image visibility toggle to map controls</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/map-toggle.tsx

<li>Added Eye/EyeOff button to toggle attached image visibility<br> <li> Button only appears if an image is attached<br> <li> Integrated with MapToggleContext and MapDataContext


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/179/files#diff-80429983cc555fc2edff59b488ff5dc258eb19568f03a5e71f2a3105302894ac">+37/-21</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>map-data-context.tsx</strong><dd><code>Add attachedImage field to map data context</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/map/map-data-context.tsx

- Added attachedImage field to MapData interface


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/179/files#diff-962700610dd525e2640fcbf6400b4f1486142becccbca3b6e0c941e42ef5af6a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>mapbox-map.tsx</strong><dd><code>Render attached image as raster layer on map</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/map/mapbox-map.tsx

<li>Added effect to render attached image as raster layer<br> <li> Handles adding/removing image sources and layers dynamically<br> <li> Respects image visibility toggle from context


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/179/files#diff-ce7ea3ef85a6812c7da39941ffa47e8e0fadbb3fc7c391cf1cafd96303cf3a0f">+64/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

google-labs-jules[bot] committed:
feat: Implement map image attachment and toggle functionality

This commit introduces the ability for you to attach an image to be displayed on the map, along with controls to manage its visibility and removal.

Key changes:

- **Image Attachment:**
    - Added a file input to `ChatPanel` allowing you to select image files.
    - Selected images are read as data URLs and stored in `MapDataContext`.
- **Map Rendering:**
    - `MapboxMap` component now consumes the attached image data from `MapDataContext`.
    - Attached images are rendered as a raster layer on the map using a default coordinate quad.
    - The map updates automatically if the image is changed or cleared.
- **Visibility Toggle:**
    - `MapToggleContext` updated to include state for image visibility (`isAttachedImageVisible`).
    - `MapToggle` component now includes a button (Eye/EyeOff icon) to toggle the attached image's visibility. This button only appears if an image is attached.
    - `MapboxMap` respects this toggle, showing or hiding the image layer without removing it from the source unless the image is cleared entirely.
- **Remove Functionality:**
    - Added a "Remove Image" button (XCircle icon) to `ChatPanel`.
    - This button clears the `attachedImage` data from `MapDataContext`, causing the image to be removed from the map.
- **Context Updates:**
    - `MapDataContext` (`map-data-context.tsx`) updated to include `attachedImage?: string | File | null;`.
    - `MapToggleContext` (`map-toggle-context.tsx`) updated to include `isAttachedImageVisible` and its setter.
- **UI Adjustments:**
    - Icons and buttons added to `ChatPanel` and `MapToggle`.
    - Padding adjusted in `ChatPanel` to accommodate new UI elements.
    - Added `title` attributes for better accessibility on new interactive elements.

This feature enhances the map's utility by allowing visual information to be overlaid directly by you.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/QueueLab/QCX?pullRequest=179) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/QueueLab/QCX?pullRequest=179) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/QueueLab/QCX?pullRequest=179) it.</sub>

google-labs-jules[bot] committed:
fix: Resolve build errors after map image attachment feature

This commit addresses build errors you encountered after the introduction of the map image attachment functionality.

The following fixes were implemented:

1.  **Corrected Module Import Paths:**
    - Adjusted the import path for `map-data-context` in `components/chat-panel.tsx` from `../map/map-data-context` to `./map/map-data-context`.
    - Adjusted the import path for `map-data-context` in `components/map-toggle.tsx` from `../map/map-data-context` to `./map/map-data-context`.

2.  **Ensured Global Context Availability:**
    - Wrapped the application content within `MapDataProvider` in `app/layout.tsx`. This resolves runtime errors during prerendering (e.g., for `/_not-found`) caused by components attempting to use `useMapData` outside of a provider.

These changes ensure that your project compiles successfully with `bun run build`.

google-labs-jules[bot] committed:
feat: Implement AI-driven image understanding for map rendering

This commit introduces functionality to analyze attached images using an AI model. If an image is identified as a map, the model extracts its geographical bounding box, and the image is then rendered at that specific location on the Mapbox map.

Key changes:

1.  **AI Model Integration (`chat-panel.tsx`):**
    - On image attachment, the image data is sent to an AI model (obtained via `getModel()` from `lib/utils`).
    - A prompt instructs the model to determine if the image is a map and, if so, to return a JSON object containing `isMap: boolean` and `boundingBox: { northEast: {lat, lon}, southWest: {lat, lon} }`.
    - Loading states and error messages are displayed in the UI during AI processing.

2.  **Updated Data Context (`map-data-context.tsx`):**
    - `MapData` interface extended with:
        - `imageBoundingBox?: BoundingBox | null;` (where `BoundingBox` is a new interface for corner coordinates).
        - `error?: string | null;` (to store errors from image processing).

3.  **Dynamic Map Rendering (`mapbox-map.tsx`):**
    - The component now consumes `imageBoundingBox` and `error` from `MapDataContext`.
    - If `mapData.attachedImage` and a valid `mapData.imageBoundingBox` are present (and no critical error), the image is rendered on the map using these dynamic coordinates. The fixed placeholder coordinates are no longer used.
    - If `imageBoundingBox` is not available or an error occurs, the image is not rendered (or is removed if previously rendered).
    - The `useEffect` hook managing image rendering has been updated to react to changes in `imageBoundingBox` and `error`.

4.  **Error Handling and UI Feedback:**
    - `chat-panel.tsx` now provides feedback to you if the AI model fails to process the image, if it's not identified as a map, or if coordinates cannot be determined.
    - The attach button shows a spinner during processing.

This significantly enhances the map attachment feature by moving from static image placement to intelligent, content-aware rendering.

google-labs-jules[bot] committed:
feat: AI image understanding for dynamic map rendering & build fixes

This commit introduces functionality to analyze attached images using an AI model. If an image is identified as a map, the model extracts its geographical bounding box, and the image is rendered at that specific location on the Mapbox map.

This version also includes critical build fixes that arose during the initial feature implementation.

Key Feature Changes:

1.  **AI Model Integration (`chat-panel.tsx`):**
    - On image attachment, image data is sent to an AI model (via `getModel()`).
    - The model is prompted to identify if the image is a map and provide a `boundingBox` in JSON format.
    - The UI now shows loading spinners and error messages during AI processing.

2.  **Updated Data Context (`map-data-context.tsx`):**
    - The `MapData` interface has been extended with `imageBoundingBox?: BoundingBox | null;` and `error?: string | null;`.

3.  **Dynamic Map Rendering (`mapbox-map.tsx`):**
    - Images identified as maps are rendered using the AI-provided `imageBoundingBox`.
    - Fallback mechanisms handle cases where a bounding box isn't available or an error occurs.

Build Fixes Implemented:

1.  **`components/chat-panel.tsx`:**
    - I resolved a string syntax error ("Expected unicode escape") in the AI prompt by converting it to a template literal.
    - I fixed a TypeScript error by removing an invalid `className` prop from the `Spinner` component.
2.  **`components/map/mapbox-map.tsx`:**
    - I corrected a TypeScript error by changing the type for image coordinates from `mapboxgl.ImageSourceOptions['coordinates']` to `mapboxgl.ImageSourceSpecification['coordinates']`.

These changes ensure the new feature is functional and the project compiles successfully.

google-labs-jules[bot] committed:
fix: Resolve all build errors for image understanding feature

This commit addresses and resolves all outstanding build errors related to the AI-driven image understanding and dynamic map rendering feature. It ensures the project compiles successfully with `bun run build`.

The following issues were addressed:

1.  **Merge Conflicts (`components/map/mapbox-map.tsx`):**
    - Ensured that any previously reported merge conflicts in this file were resolved, integrating feature changes with updates from the main branch.

2.  **Syntax Error (`components/map/map-data-context.tsx`):**
    - Resolved a persistent syntax error (`Expected '{', got 'interface'`) by rewriting the file content, which likely cleared hidden problematic characters or encoding issues.

3.  **TypeScript and Syntax Errors from Previous Iterations:**
    - Corrected a string syntax error ("Expected unicode escape") in the AI prompt in `components/chat-panel.tsx` by converting it to a template literal.
    - Fixed a TypeScript error in `components/chat-panel.tsx` by removing an invalid `className` prop from the `Spinner` component.
    - Corrected a TypeScript error in `components/map/mapbox-map.tsx` by changing the type for image coordinates from `mapboxgl.ImageSourceOptions['coordinates']` to `mapboxgl.ImageSourceSpecification['coordinates']`.
    - Ensured `MapDataProvider` is correctly wrapping the application in `app/layout.tsx` to prevent context-related errors during prerendering.
    - Corrected various import paths for contexts and components as needed.

All features related to image attachment, AI model processing for map coordinate extraction, and dynamic rendering on Mapbox, along with all necessary build and type error corrections, are included. The application should now be stable and buildable.

ngoiyaeric commented:
@coderabbitai full review
